service/frontend/workflowHandler.go

```diff
@@ -1622,6 +1622,24 @@ func (wh *WorkflowHandler) RespondActivityTaskFailed(
 	sizeLimitError := wh.config.BlobSizeLimitError(namespaceEntry.Name().String())
 	sizeLimitWarn := wh.config.BlobSizeLimitWarn(namespaceEntry.Name().String())
 
+	// Record details as heartbeat if present
+	if request.GetLastHeartbeatDetails() != nil {
+		response, err := wh.RecordActivityTaskHeartbeat(ctx, &workflowservice.RecordActivityTaskHeartbeatRequest{
```

Comment:
1. That is the intention.
2. I see your point and makes sense. 
 
The way the SDK intends to use this is when an activity fails, as apart of flushing state if there is also heartbeat event not yet, the API requests could be combined. This would imply there would be no other heartbeats and this is the final call to the activity task.

Today the SDK and client do not make use of this, but your statement above means that ordering could be off if another heartbeat API request comes in between the transactions. This would mean likely some ordering violation on the SDK side, and nothing in this PR would prevent that from happening.

Ideally the goal was to appear as a convenience wrapper for the case of clean up.
There was an edge case where the details blob is larger than allowed and this path let that logic simply be reused.

Let me dig a little deeper on this.

Reply:
Frontend should pass all to history in one RPC call. You need to update history service's proto to include that extra heartbeat field as well. History service should make one transaction update that include both heartbeat and fail activity.

Reply:
I have updated to pass all to history.

I could use some eyes on my tests please.